core/src/main/java/org/apache/calcite/rel/rules/JoinDeriveEquivalenceFilterRule.java

```diff
@@ -0,0 +1,380 @@
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+
+import org.immutables.value.Value;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ *  Planner rule that derives more equivalent predicates from inner
+ *  {@link Join} and creates {@link Filter} with those predicates.
+ *  Then {@link FilterJoinRule} will try to push these new predicates down.
+ *  (So if you enable this rule, please make sure to enable {@link FilterJoinRule} also).
+ *  <p>The basic idea is that, for example, in the query
+ *  <blockquote>SELECT * FROM ta INNER JOIN tb ON ta.x = tb.y WHERE ta.x &gt; 10</blockquote>
+ *  we can infer condition tb.y &gt; 10 and push it down to the table tb.
+ *  In this way, maybe we can reduce the amount of data involved in the {@link Join}.
+ *  <p>For example, the query plan
+ *  <blockquote><pre>{@code
+ *  LogicalJoin(condition=[=($1, $5)], joinType=[inner])
+ *   LogicalTableScan(table=[[hr, emps]])
+ *   LogicalFilter(condition=[>($0, 10)])
+ *     LogicalTableScan(table=[[hr, depts]])
+ *  }</pre></blockquote>
+ *  <p> will convert to
+ *  <blockquote><pre>{@code
+ *  LogicalJoin(condition=[=($1, $5)], joinType=[inner])
+ *   LogicalFilter(condition=[>($1, 20)])
+ *     LogicalTableScan(table=[[hr, emps]])
+ *   LogicalFilter(condition=[>($0, 20)])
+ *     LogicalTableScan(table=[[hr, depts]])
+ *  }</pre></blockquote>
+ *  <p>the query plan
+ *  <blockquote><pre>{@code
+ *  LogicalJoin(condition=[=($1, $5)], joinType=[inner])
+ *   LogicalFilter(condition=[SEARCH($1, Sarg[(10..30)])])
+ *     LogicalTableScan(table=[[hr, emps]])
+ *   LogicalFilter(condition=[SEARCH($0, Sarg[(20..40)])])
+ *     LogicalTableScan(table=[[hr, depts]])
+ *  }</pre></blockquote>
+ *  <p> will convert to
+ *  <blockquote><pre>{@code
+ *  LogicalJoin(condition=[=($1, $5)], joinType=[inner])
+ *   LogicalFilter(condition=[SEARCH($1, Sarg[(20..30)])])
+ *     LogicalTableScan(table=[[hr, emps]])
+ *   LogicalFilter(condition=[SEARCH($0, Sarg[(20..30)])])
+ *     LogicalTableScan(table=[[hr, depts]])
+ *  }</pre></blockquote>
+ *  <p>Currently, the rule has some limitations:
+ *  <p>1. only handle partial predicates (comparison), but this can be extended to
+ *      support more predicates such as 'LIKE', 'RLIKE' and 'SIMILAR' in the future.
+ *  <p>2. only support simple condition inference, such as: {$1 = $2, $2 = 10} =&gt; {$1 = 10},
+ *     can not handle complex condition inference, such as conditions with functions, like
+ *     {a = b, b = abs(c), c = 1} =&gt; {a = abs(1)}
+ *  <p>3. only support discomposed literal, for example
+ *     it can infer {$1 = $2, $1 = 10} =&gt; {$2 = 10}
+ *     it can not infer {$1 = $2, $1 = 10 + 10} =&gt; {$2 = 10 + 10}
+ */
+
+@Value.Enclosing
+public class JoinDeriveEquivalenceFilterRule
```

Comment:
This name is misleading since it is not just equivalency, no?

Reply:
Yes, this is indeed a bit misleading. I originally wanted to express the derivation of rules based on equivalent rewriting. I will rename it.